test/test_multidevice_pipeline.cpp

```diff
@@ -596,59 +596,88 @@ TEST_F(PipelineTest, matmuls_megatron_mlp) {
 // TODO: Temporary hack of adding a zero bias so that global output buffer
 // is properly allocated.
 TEST_F(PipelineTest, matmul_megatron_attention) {
-  // ignoring embarassingly parallel attention heads
   FusionGuard fg(fusion.get());
   int num_devices = communicator->size();
   std::vector<int64_t> devices(num_devices);
   std::iota(devices.begin(), devices.end(), 0);
   DeviceMesh mesh(devices);
 
-  int H = 4;
-  int S = 8;
-  int D_model = num_devices;
-  int D = 3;
+  int H = num_devices; // number of attention heads
+  int S = 8; // sequence length
+  int D_model = 5;
+  int Dv = 4;
+  int Dk = 4;
+
+  TensorView* q = makeConcreteTensor({H, S, Dk}); //64
+  TensorView* k = makeConcreteTensor({H, S, Dk}); //64
+  TensorView* v = makeConcreteTensor({H, S, Dv});  //64
+  TensorView* w = makeConcreteTensor({H, Dv, D_model}); //40
+  TensorView* bias = makeConcreteTensor({S, D_model});//40
+  auto tv_inputs = {q, k, v, w, bias};
+  for (auto i : tv_inputs) {
+    fusion->addInput(i);
+  }
 
-  TensorView* tvx = makeConcreteTensor({S, H*D});  // (S,H*D)
-  TensorView* tvw = makeConcreteTensor({D_model, H*D});  // w^T (D_model, H*D) 
-  TensorView* bias = makeConcreteTensor({D_model, S}); 
-  fusion->addInput(tvx);
-  fusion->addInput(tvw);
-  fusion->addInput(bias);
+  // scaled dot product attention
+  TensorView* q_b = broadcast(q, {false, false, true, false}); // (H, S, S, Dk)
+  TensorView* k_b = broadcast(k, {false, true, false, false}); // (H, S, S, Dk)
+  TensorView* qk_ = mul(q_b, k_b); // (H, S, S, Dk)
+  TensorView* qk = sum(qk_, {3});  // (H, S, S)
+
+  TensorView* qk_b = broadcast(qk, {false, false, false, true}); // (H, S, S, Dv)
+  TensorView* v_b = broadcast(v, {false, true, false, false}); // (H, S, S, Dv)
+  TensorView* qkv_ = mul(qk_b, v_b);
+  TensorView* qkv = sum(qkv_, {2}); // (H, S, Dv)
+
+  // linear
+  TensorView* qkv_b = broadcast(qkv, {false, false, false, true}); // (H, S, Dv, Dmodel)
+  TensorView* w_b = broadcast(w, {false, true, false, false}); // (H, S, Dv, Dmodel)
+  TensorView* y0 = mul(qkv_b, w_b); 
+  TensorView* y0_ = set(y0); // Temporarily force y0 to be fully replicated
```

Comment:
you mean fully sharded right?
It should be sharded over the H dimension. There I don't think we need this `y0_` at all

Reply:
good catch we don't need y0_